python/tvm/relay/frontend/pytorch.py

```diff
@@ -0,0 +1,974 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=import-self, too-many-lines, len-as-condition, no-else-return, unused-variable, too-many-nested-blocks
+# pylint: disable=consider-iterating-dictionary, invalid-name, unused-argument, unused-variable
+"""PT: PyTorch frontend."""
+import numpy as np
+
+import tvm
+
+from .. import expr as _expr
+from .. import module as _module
+from .. import op as _op
+from .common import get_relay_op
+from .common import infer_type as _infer_type
+
+__all__ = ['from_pytorch']
+
+# operator implementation
+def _elemwise(name):
+    def _impl(inputs):
+        data0 = convert_input(inputs[0])
+        data1 = convert_input(inputs[1])
+
+        if not isinstance(data0, (_expr.Call, _expr.TupleGetItem, _expr.Var)):
+            temp = data0
+            data0 = data1
+            data1 = temp
+
+        return get_relay_op(name)(data0, data1)
+    return _impl
+
+
+def _unsqueeze():
+    def _impl(inputs):
+        data = inputs[0]
+        axis = inputs[1]
+
+        return _op.transform.expand_dims(data, int(axis), 1)
+    return _impl
+
+def _concatenate():
+    def _impl(inputs):
+        data = inputs[0]
+        axis = inputs[1]
+
+        if isinstance(data, (_expr.Call, _expr.TupleGetItem, _expr.Var)):
+            data = [data]
+
+        return _op.tensor.concatenate(data, int(axis))
+    return _impl
+
+def _slice():
+    def _impl(inputs):
+        data = inputs[0]
+        strides = []
+
+        inferred_shape = _infer_type(data).checked_type.shape
+        end = []
+        for infer in inferred_shape:
+            end.append(int(infer))
+        if isinstance(data, _expr.Var):
+            end = get_tensor_from_relay_var(data)
+
+        begin = [0]*len(end)
+        dim = int(inputs[1])
+        begin[dim] = int(inputs[2])
+
+        if len(inputs[3]) < 10:
+            end[dim] = min(end[dim], int(inputs[3]))
+
+        strides.append(int(inputs[4]))
+        return _op.transform.strided_slice(data, begin, end, strides)
+    return _impl
+
+def _select():
+    def _impl(inputs):
+        data = inputs[0]
+        inferred_shape = _infer_type(data).checked_type.shape
+        end = []
+
+        for infer in inferred_shape:
+            end.append(int(infer))
+
+        begin = [0]*len(end)
+        dim = int(inputs[1])
+        index = int(inputs[2])
+
+        end[dim] = index+1
+        begin[dim] = index
+
+        strides = [1]
+
+        sym = _op.transform.strided_slice(data, begin, end, strides)
+        axis = [dim]
+
+        return _op.transform.squeeze(sym, axis)
+    return _impl
+
+def _ones():
+    def _impl(inputs):
+        fill_value = _expr.const(1, dtype='float32')
```

Comment:
we might need to consider different precision in the future

Reply:
Yes I agree. I think because DLR only supports float32, this is fine for now (upstream changes will have different types once I figure out PyTorch typing).